### ISSUE TITLE
Add support for missing data in discrete-time models

### ISSUE THREAD

**Issue by dimkab** [BasisResearch/dynestyx#138]
Modify `DiscreteTimeSimulator::_simulate` to support `NaN`s in some observation components.
- [ ] l410: adjust the condition to take partial NaNs into account (i.e. enter the block only if all components are NOT NaNs)
- [ ] l484, l490: adjust the obs for DiracObservations
- [ ] add utilities to factor multivariate Gaussians [optional, another PR?]

Prefer no dispatches :)

Tests: TBD

**Comment by DanWaxman** [comment]
Is there anything left here after #239 #234 (cc @mattlevine22)? I suppose we need to get things working with continuous-time models, but that seems more on CD-Dynamax and/or discretizer + discrete-time filter/smoother.

**Comment by mattlevine22** [comment]
Yeah I wanna do DiracIdentity next...it should probably handle full dropout at least. Could open separate issue for that if you want. @dimkab how are you handling things in your custom numpyro code these days...I'd really like to support what you're doing via dynestyx settings if possible.

**Comment by dimkab** [comment]
@mattlevine22 essentially as in my PR:
https://github.com/BasisResearch/collab-dynmodels/blob/55b250699d7881998a0633fb8abb6d144924b5aa/src/inference_models/likelihood_engines.py#L278

**Comment by mattlevine22** [comment]
Thanks for linking @dimkab !

TLDR: 
1. Have you considered using `LinearGaussianObservation` instead? Might adequately mimic what you have already. Happy to talk about what might be different.
2. At the end, I propose some code for how to vectorize things more broadly (something dynestyx and I think your collab-dynmodels similarly struggle with).

Details...
If I understand your PR correctly, you use `numpyro.contrib.control_flow.scan` to step through transitions in the general missingness case, right? If so, I think our main Simulator w/ non-Dirac interface (e.g., create a LinearGaussianObservation with a desired noise level) does what you need. 

The main benefit of DiracIdentity stuff is to`vmap` instead of scan. But once there is missingness, numpyro really wants us to do a scan...I think your above code reflects this challenge. Seems like the one case where your code does vectorized stuff is in `skip=True` mode where all measurements were dropped in a block (not partially-missing); we didn't implement this specific idea for Simulator+Dirac, but we could if its important.

More generally (curious about @DanWaxman 's take), we could do a lot more vectorization under missingness if we let go of the PPL beauty under-the-hood. Something like:

```python

x0 = numpyro.sample("f_x0", dynamics.initial_condition)
# create a dummy distribution for the latents---we will subtract off its log-prob later.
z = numpyro.sample("f_x_latents", dist.Normal(0, 1).expand((T - 1, state_dim)).to_event(2))

x_full = jnp.concatenate([x0[None], z], axis=0)

x_prev = x_full[:-1]
x_next = x_full[1:]

# transition log-probs over time; 
# this can be trivially vmapped over (x_prev, x_next, times[:-1], times[1:])
# subtract off log-prob(z)
numpyro.factor(
    "f_latent_lp",
    sum_t log p(x_next[t] | x_prev[t]) - log q(z), 
)

# observation log-probs over time; 
# can hopefully prep this for vmapping even under missingness
numpyro.factor(
    "f_obs_lp",
    sum_t log p(y_observed_t | x_full[t]),
)
```

Note that this is NOT a simulator (it is not a generative process for data), but it IS a simple strategy for joint inference of latent states and model parameters (which allows us to cope with missingness).

The choice of Gaussian for x_latent should only really impact us if we don't explicitly choose an initialization for the latent states (the gaussian log prob gets subtracted off; it is just gaussian to create the sites). Could also do ImproperUniform or something.

**Comment by dimkab** [comment]
@mattlevine22 I think what you suggested makes sense for the general case. For our use case of  T,N large and DiracObservation, it may bite us to add artificial small observation noise at the price of  having the full latents and possibly biased estimates (the filtering would only be an approximation anyway...)
I like the vectorization idea - it is even easier to implement for the DiracObservations; I'll try to do that soon in collab-dynmodels and let you know how it goes there.

**Comment by mattlevine22** [comment]
Yeah, so my high-level point is that the vectorization idea can be orthogonal to the noise assumptions.
- For DiracObservation (trust all measurements fully), the number of latent variables = number of missing observations (hopefully kinda small)
- For NoisyObservations, the number of latent variables = T x state_dim (possibly quite large)

Will be very curious to see how the vectorization stuff works out for you, happy to discuss further and/or develop it into dynestyx.
